File: baml_language/sdk_tests/crates/cpp/type_shapes/customizable/tests/test_generics.cc
// Roundtrip coverage for the generics suite -- generic classes over <int>.
// Port of type_shapes/customizable/roundtrip_tests/test_generics.py. The
// generic instance-method path is covered separately in test_generic.cpp;
// here we cover the concretely-instantiated generic class round trips.
//
// Recursive generic references are boxed in C++ (GenericLinkedList.next is
// OptionalBox, NestedGenerics.wr wraps Box<GenericLinkedList>), so the
// construction shapes differ from Python only in those Box spellings.
#include <baml_sdk.h>
#include <baml_test.h>

#include <cstdint>
#include <optional>
#include <vector>

namespace generics = baml_sdk::generics;
using generics::GenericBinaryTree;
using generics::GenericLinkedList;
using generics::Wrapper;

BAML_TEST(round_trip_wrapper_int) {
  const Wrapper<int64_t> w{5};
  BAML_ASSERT(generics::round_trip_wrapper_int(w) == w);
}

BAML_TEST(round_trip_generic_linked_list_int) {
  const GenericLinkedList<int64_t> ll{
      1, GenericLinkedList<int64_t>{
             2, ::baml::OptionalBox<GenericLinkedList<int64_t>>()}};
  BAML_ASSERT(generics::round_trip_generic_linked_list_int(ll) == ll);
}

BAML_TEST(round_trip_generic_binary_tree_int) {
  const GenericBinaryTree<int64_t> t{
      1, ::baml::OptionalBox<GenericBinaryTree<int64_t>>(),
      ::baml::OptionalBox<GenericBinaryTree<int64_t>>()};
  BAML_ASSERT(generics::round_trip_generic_binary_tree_int(t) == t);
}

BAML_TEST(round_trip_box_int) {
  const generics::Box<int64_t> b{3, Wrapper<int64_t>{4}};
  BAML_ASSERT(generics::round_trip_box_int(b) == b);
}

BAML_TEST(round_trip_nested_generics) {
  const generics::NestedGenerics n{
      Wrapper<Wrapper<int64_t>>{Wrapper<int64_t>{1}},
      Wrapper<std::vector<int64_t>>{{1, 2}},
      Wrapper<::baml::Box<GenericLinkedList<int64_t>>>{
          GenericLinkedList<int64_t>{
              9, ::baml::OptionalBox<GenericLinkedList<int64_t>>()}},
  };
  BAML_ASSERT(generics::round_trip_nested_generics(n) == n);
}

BAML_TEST(round_trip_differing_instantiation) {
  const generics::DifferingInstantiation d{GenericLinkedList<Wrapper<int64_t>>{
      Wrapper<int64_t>{1},
      ::baml::OptionalBox<GenericLinkedList<Wrapper<int64_t>>>()}};
  BAML_ASSERT(generics::round_trip_differing_instantiation(d) == d);
}

File: baml_language/sdk_tests/crates/python_pydantic2/type_shapes/customizable/roundtrip_tests/test_generics.py
"""Roundtrip coverage for `baml_sdk.generics` — generic classes (over `<int>`).

The generic *instance method* path (`WrapperMethods.get_value` /
`get_value_or_marker`) is covered separately in
`customizable/test_generic.py`; here we cover the
concretely-instantiated generic class round trips.
"""

import baml_sdk  # noqa: F401  — initializes the BAML runtime
from baml_sdk.generics import (
    Wrapper,
    GenericLinkedList,
    GenericBinaryTree,
    Box,
    NestedGenerics,
    DifferingInstantiation,
    round_trip_wrapper_int,
    round_trip_generic_linked_list_int,
    round_trip_generic_binary_tree_int,
    round_trip_box_int,
    round_trip_nested_generics,
    round_trip_differing_instantiation,
)


def test_generics_round_trip_wrapper_int():
    w = Wrapper[int](value=5)
    assert round_trip_wrapper_int(w=w) == w


def test_generics_round_trip_generic_linked_list_int():
    ll = GenericLinkedList[int](value=1, next=GenericLinkedList[int](value=2, next=None))
    assert round_trip_generic_linked_list_int(l=ll) == ll


def test_generics_round_trip_generic_binary_tree_int():
    t = GenericBinaryTree[int](value=1, left=None, right=None)
    assert round_trip_generic_binary_tree_int(t=t) == t


def test_generics_round_trip_box_int():
    b = Box[int](value=3, wrapped=Wrapper[int](value=4))
    assert round_trip_box_int(b=b) == b


def test_generics_round_trip_nested_generics():
    n = NestedGenerics(
        ww=Wrapper[Wrapper[int]](value=Wrapper[int](value=1)),
        wl=Wrapper[list](value=[1, 2]),
        wr=Wrapper[GenericLinkedList[int]](
            value=GenericLinkedList[int](value=9, next=None)
        ),
    )
    assert round_trip_nested_generics(n=n) == n


def test_generics_round_trip_differing_instantiation():
    d = DifferingInstantiation(
        list=GenericLinkedList[Wrapper[int]](value=Wrapper[int](value=1), next=None)
    )
    assert round_trip_differing_instantiation(d=d) == d
